MacOS wheel build fails when installing NumPy

### What happened?

[mac-arm-ventura-unprovisioned-clang-wheel-nightly-release](https://drake-jenkins.csail.mit.edu/job/mac-arm-ventura-unprovisioned-clang-wheel-nightly-release/287/) failed when building Numpy 2.1.0. See output log below.

#21829 proposed a temporary fix to upper bound the mac binary Numpy version to 2.0.1.

### Version

_No response_

### What operating system are you using?

macOS 13 (Ventura)

### What installation option are you using?

_No response_

### Relevant log output

```shell
[2024-08-19T20:40:48.053Z]   Preparing metadata (pyproject.toml): started
[2024-08-19T20:41:00.963Z]   Preparing metadata (pyproject.toml): finished with status 'error'
[2024-08-19T20:41:00.963Z]   error: subprocess-exited-with-error
[2024-08-19T20:41:00.963Z]   
[2024-08-19T20:41:00.963Z]   × Preparing metadata (pyproject.toml) did not run successfully.
[2024-08-19T20:41:00.963Z]   │ exit code: 1
[2024-08-19T20:41:00.963Z]   ╰─> [292 lines of output]
[2024-08-19T20:41:00.963Z]       + /opt/drake-wheel-build/python3.11/python/bin/python3.11 /private/var/folders/0h/x72zp8qj45xfftcw9yh2hg_m0000gn/T/pip-install-zmukoya7/numpy_0fcae2117b8943b79311589194dfcdfe/vendored-meson/meson/meson.py setup /private/var/folders/0h/x72zp8qj45xfftcw9yh2hg_m0000gn/T/pip-install-zmukoya7/numpy_0fcae2117b8943b79311589194dfcdfe /private/var/folders/0h/x72zp8qj45xfftcw9yh2hg_m0000gn/T/pip-install-zmukoya7/numpy_0fcae2117b8943b79311589194dfcdfe/.mesonpy-75i1sxqh -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md --native-file=/private/var/folders/0h/x72zp8qj45xfftcw9yh2hg_m0000gn/T/pip-install-zmukoya7/numpy_0fcae2117b8943b79311589194dfcdfe/.mesonpy-75i1sxqh/meson-python-native-file.ini
[2024-08-19T20:41:00.963Z]       The Meson build system
[2024-08-19T20:41:00.963Z]       Version: 1.4.99
[2024-08-19T20:41:00.963Z]       Source dir: /private/var/folders/0h/x72zp8qj45xfftcw9yh2hg_m0000gn/T/pip-install-zmukoya7/numpy_0fcae2117b8943b79311589194dfcdfe
[2024-08-19T20:41:00.963Z]       Build dir: /private/var/folders/0h/x72zp8qj45xfftcw9yh2hg_m0000gn/T/pip-install-zmukoya7/numpy_0fcae2117b8943b79311589194dfcdfe/.mesonpy-75i1sxqh
[2024-08-19T20:41:00.963Z]       Build type: native build
[2024-08-19T20:41:00.963Z]       Project name: NumPy
[2024-08-19T20:41:00.963Z]       Project version: 2.1.0
[2024-08-19T20:41:00.963Z]       C compiler for the host machine: cc (clang 14.0.3 "Apple clang version 14.0.3 (clang-1403.0.22.14.1)")
[2024-08-19T20:41:00.963Z]       C linker for the host machine: cc ld64 857.1
[2024-08-19T20:41:00.963Z]       C++ compiler for the host machine: c++ (clang 14.0.3 "Apple clang version 14.0.3 (clang-1403.0.22.14.1)")
[2024-08-19T20:41:00.963Z]       C++ linker for the host machine: c++ ld64 857.1
[2024-08-19T20:41:00.963Z]       Cython compiler for the host machine: cython (cython 3.0.11)
[2024-08-19T20:41:00.963Z]       Host machine cpu family: aarch64
[2024-08-19T20:41:00.963Z]       Host machine cpu: aarch64
[2024-08-19T20:41:00.963Z]       Program python found: YES (/opt/drake-wheel-build/python3.11/python/bin/python3.11)
[2024-08-19T20:41:00.963Z]       Found pkg-config: YES (/opt/homebrew/bin/pkg-config) 0.29.2
[2024-08-19T20:41:00.963Z]       Run-time dependency python found: YES 3.11
[2024-08-19T20:41:00.963Z]       Has header "Python.h" with dependency python-3.11: YES
[2024-08-19T20:41:00.963Z]       Compiler for C supports arguments -fno-strict-aliasing: YES
[2024-08-19T20:41:00.963Z]       Compiler for C supports arguments -ftrapping-math: YES
[2024-08-19T20:41:00.963Z]       Compiler for C supports link arguments -Wl,-ld_classic: NO
[2024-08-19T20:41:00.963Z]       Message: During parsing cpu-dispatch: The following CPU features were ignored due to platform incompatibility or lack of support:
[2024-08-19T20:41:00.963Z]       "XOP FMA4"
[2024-08-19T20:41:00.963Z]       Test features "NEON NEON_FP16 NEON_VFPV4 ASIMD" : Supported
[2024-08-19T20:41:00.963Z]       Test features "ASIMDHP" : Supported
[2024-08-19T20:41:00.963Z]       Test features "ASIMDFHM" : Supported
[2024-08-19T20:41:00.963Z]       Test features "SVE" : Unsupported due to Compiler fails against the test code of "SVE"
[2024-08-19T20:41:00.963Z]       Configuring npy_cpu_dispatch_config.h using configuration
[2024-08-19T20:41:00.963Z]       Message:
[2024-08-19T20:41:00.963Z]       CPU Optimization Options
[2024-08-19T20:41:00.963Z]         baseline:
[2024-08-19T20:41:00.963Z]           Requested : min
[2024-08-19T20:41:00.963Z]           Enabled   : NEON NEON_FP16 NEON_VFPV4 ASIMD
[2024-08-19T20:41:00.963Z]         dispatch:
[2024-08-19T20:41:00.963Z]           Requested : max -xop -fma4
[2024-08-19T20:41:00.963Z]           Enabled   : ASIMDHP ASIMDFHM
[2024-08-19T20:41:00.963Z]       
[2024-08-19T20:41:00.963Z]       Library m found: YES
[2024-08-19T20:41:00.963Z]       Run-time dependency scipy-openblas found: NO (tried pkgconfig)
[2024-08-19T20:41:00.963Z]       Run-time dependency accelerate found: NO (tried system)
[2024-08-19T20:41:00.963Z]       Found CMake: /opt/homebrew/bin/cmake (3.30.2)
[2024-08-19T20:41:00.964Z]       WARNING: CMake Toolchain: Failed to determine CMake compilers state
[2024-08-19T20:41:00.964Z]       Run-time dependency openblas found: NO (tried pkgconfig, pkgconfig, pkgconfig, system and cmake)
[2024-08-19T20:41:00.964Z]       Run-time dependency flexiblas found: NO (tried pkgconfig, framework and cmake)
[2024-08-19T20:41:00.964Z]       Run-time dependency blis found: NO (tried pkgconfig, framework and cmake)
[2024-08-19T20:41:00.964Z]       Run-time dependency blas found: YES unknown
[2024-08-19T20:41:00.964Z]       ../numpy/meson.build:135: WARNING: Project targets '>=1.2.99' but uses feature introduced in '1.3.0': dep 'blas' custom lookup.
[2024-08-19T20:41:00.964Z]       Message: BLAS symbol suffix:
[2024-08-19T20:41:00.964Z]       Run-time dependency accelerate found: NO (tried system)
[2024-08-19T20:41:00.964Z]       Run-time dependency openblas found: NO (tried pkgconfig, pkgconfig, pkgconfig, system and cmake)
[2024-08-19T20:41:00.964Z]       Run-time dependency flexiblas found: NO (tried pkgconfig, framework and cmake)
[2024-08-19T20:41:00.964Z]       Run-time dependency lapack found: YES unknown
[2024-08-19T20:41:00.964Z]       ../numpy/meson.build:198: WARNING: Project targets '>=1.2.99' but uses feature introduced in '1.3.0': dep 'lapack' custom lookup.
[2024-08-19T20:41:00.964Z]       Checking if "Check atomic builtins without -latomic" : links: NO
[2024-08-19T20:41:00.964Z]       Library atomic found: NO
[2024-08-19T20:41:00.964Z]       Program _build_utils/process_src_template.py found: YES (/opt/drake-wheel-build/python3.11/python/bin/python3.11 /private/var/folders/0h/x72zp8qj45xfftcw9yh2hg_m0000gn/T/pip-install-zmukoya7/numpy_0fcae2117b8943b79311589194dfcdfe/numpy/_build_utils/process_src_template.py)
[2024-08-19T20:41:00.964Z]       Program _build_utils/tempita.py found: YES (/opt/drake-wheel-build/python3.11/python/bin/python3.11 /private/var/folders/0h/x72zp8qj45xfftcw9yh2hg_m0000gn/T/pip-install-zmukoya7/numpy_0fcae2117b8943b79311589194dfcdfe/numpy/_build_utils/tempita.py)

...

[2024-08-19T20:41:00.966Z]       Build targets in project: 63
[2024-08-19T20:41:00.966Z]       WARNING: Project specifies a minimum meson_version '>=1.2.99' but uses features which were added in newer versions:
[2024-08-19T20:41:00.966Z]        * 1.3.0: {'dep 'blas' custom lookup', 'dep 'lapack' custom lookup'}
[2024-08-19T20:41:00.966Z]       
[2024-08-19T20:41:00.966Z]       NumPy 2.1.0
[2024-08-19T20:41:00.966Z]       
[2024-08-19T20:41:00.966Z]         User defined options
[2024-08-19T20:41:00.966Z]           Native files: /private/var/folders/0h/x72zp8qj45xfftcw9yh2hg_m0000gn/T/pip-install-zmukoya7/numpy_0fcae2117b8943b79311589194dfcdfe/.mesonpy-75i1sxqh/meson-python-native-file.ini
[2024-08-19T20:41:00.966Z]           buildtype   : release
[2024-08-19T20:41:00.966Z]           b_ndebug    : if-release
[2024-08-19T20:41:00.966Z]           b_vscrt     : md
[2024-08-19T20:41:00.966Z]       
[2024-08-19T20:41:00.966Z]       Found ninja-1.12.1 at /opt/homebrew/bin/ninja
[2024-08-19T20:41:00.966Z]       + /opt/homebrew/bin/ninja
[2024-08-19T20:41:00.966Z]       [1/320] Generating 'numpy/_core/libnpymath.a.p/ieee754.c'
[2024-08-19T20:41:00.966Z]       [2/320] Generating 'numpy/_core/libargfunc.dispatch.h_baseline.a.p/npy_sort.h'
[2024-08-19T20:41:00.966Z]       [3/320] Generating 'numpy/_core/libargfunc.dispatch.h_baseline.a.p/arraytypes.h'
[2024-08-19T20:41:00.966Z]       [4/320] Generating numpy/_core/__ufunc_api with a custom command
[2024-08-19T20:41:00.966Z]       [31mFAILED: [0mnumpy/_core/__ufunc_api.c numpy/_core/__ufunc_api.h
[2024-08-19T20:41:00.966Z]       /opt/drake-wheel-build/python3.11/python/bin/python3.11 ../numpy/_core/code_generators/generate_ufunc_api.py -o numpy/_core
[2024-08-19T20:41:00.966Z]       Traceback (most recent call last):
[2024-08-19T20:41:00.966Z]         File "/private/var/folders/0h/x72zp8qj45xfftcw9yh2hg_m0000gn/T/pip-install-zmukoya7/numpy_0fcae2117b8943b79311589194dfcdfe/.mesonpy-75i1sxqh/../numpy/_core/code_generators/generate_ufunc_api.py", line 4, in <module>
[2024-08-19T20:41:00.966Z]           import genapi
[2024-08-19T20:41:00.966Z]       ModuleNotFoundError: No module named 'genapi'
[2024-08-19T20:41:00.966Z]       [5/320] Generating numpy/_core/npy_math_internal.h with a custom command
[2024-08-19T20:41:00.966Z]       [6/320] Generating 'numpy/_core/libnpymath.a.p/npy_math_complex.c'
[2024-08-19T20:41:00.966Z]       [7/320] Copying file numpy/__init__.py
[2024-08-19T20:41:00.966Z]       [8/320] Copying file numpy/__init__.pxd
[2024-08-19T20:41:00.966Z]       [9/320] Copying file numpy/__init__.cython-30.pxd
[2024-08-19T20:41:00.966Z]       [10/320] Generating 'numpy/_core/libhighway_qsort.dispatch.h_ASIMD.a.p/npy_sort.h'
[2024-08-19T20:41:00.966Z]       [11/320] Generating 'numpy/_core/libhighway_qsort.dispatch.h_ASIMD.a.p/arraytypes.h'
[2024-08-19T20:41:00.966Z]       [12/320] Generating 'numpy/_core/libargfunc.dispatch.h_baseline.a.p/argfunc.dispatch.c'
[2024-08-19T20:41:00.966Z]       [13/320] Generating numpy/_core/__umath_generated with a custom command
[2024-08-19T20:41:00.966Z]       ninja: build stopped: subcommand failed.
[2024-08-19T20:41:00.966Z]       [end of output]
[2024-08-19T20:41:00.966Z]   
[2024-08-19T20:41:00.966Z]   note: This error originates from a subprocess, and is likely not a problem with pip.
[2024-08-19T20:41:00.966Z] error: metadata-generation-failed
[2024-08-19T20:41:00.966Z] 
[2024-08-19T20:41:00.966Z] × Encountered error while generating package metadata.
[2024-08-19T20:41:00.966Z] ╰─> See above for output.
[2024-08-19T20:41:00.966Z] 
[2024-08-19T20:41:00.966Z] note: This is an issue with the package mentioned above, not pip.
[2024-08-19T20:41:00.966Z] hint: See above for details.
```
```

--- a/tools/wheel/wheel_builder/macos.py
+++ b/tools/wheel/wheel_builder/macos.py
@@ -21,9 +21,7 @@ _files_to_remove = []
 # default, all targets are built, but the user may down-select from this set.
 # On macOS (unlike Linux), this is just the set of Python versions targeted.
 python_targets = (
-    # TODO(jwnimmer-tri) Python 3.11 is disabled until NumPy ships PyPI wheels
-    # for that Python version that are installable on macOS 13.
-    # PythonTarget(3, 11),
+    PythonTarget(3, 11),
     PythonTarget(3, 12),
 )
 

